### ISSUE TITLE
[FIX-016] Preserve frozen/trainable parameter roles when replacing existing layers

### ISSUE THREAD

**Issue by thekaveh** [thekaveh/NNx#226]
**Category:** Model transformation integrity · **Priority:** P2 · **Size:** S

**Evidence status:** Reproduced state change; preservation policy to make explicit.

## Outcome and scope

Replacements for existing Linear weights/biases must preserve their separate trainability and module modes. Cover both widened target/downstream layers and the two SVD factors. Do not impose a new policy on genuinely new deepen layers or change function-preservation tolerances.

## Current behavior / evidence

**Context and reproduction:** Freeze every parameter of `Sequential(Linear(4,4), ReLU(), Linear(4,2))` and call `.eval()`. `widen(..., new_width=7)` returns four trainable parameter tensors, where the original had zero. `low_rank_factorize(frozen_linear, rank=4)` returns three trainable tensors. The new Linear children also default back to training mode. An optimizer constructed after surgery now updates previously frozen model content.

## Repository map

- [`src/nnx/surgery/widen.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/surgery/widen.py#L102) — Target replacement currently defaults to trainable/train mode.
- [`src/nnx/surgery/widen.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/surgery/widen.py#L138) — The downstream consumer must preserve its own independent flags.
- [`src/nnx/surgery/low_rank.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/surgery/low_rank.py#L93) — Both factors represent one source weight; the second factor carries the original bias.
- [`tests/test_surgery_widen.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_surgery_widen.py#L47) — Ownership and output parity regression foundation.
- [`tests/test_surgery_low_rank.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_surgery_low_rank.py#L14) — Full-rank numerical oracle, bias and source-independence fixtures.

## Implementation approach

1. In `widen`, propagate the target’s weight `requires_grad` to its replacement weight and target bias’s flag to its replacement bias. Independently do the same for the downstream Linear; do not copy the target’s flags onto its consumer. Set each replacement’s training flag from the replaced module, preserving mixed child modes in the copied root.
2. In `low_rank_factorize`, both `down.weight` and `up.weight` inherit `linear.weight.requires_grad`. Only `up.bias`, when present, inherits `linear.bias.requires_grad`; `down.bias` remains absent. Set the new Sequential and both factor modules to the original Linear’s mode.
3. Preserve no-source-mutation and independent tensor storage. Apply flags to freshly built replacements after copying values, not to the caller’s Parameters. Keep existing skip_init/local-generator behavior, dtype/device placement and SVD rank/error checks unchanged.
4. Show the practical optimization effect: construct the optimizer only after transformation and inspect trainable parameters; frozen content cannot enter an update because replacement constructors defaulted to trainable. Preserve valid backward behavior for trainable replacements without requiring gradients on intentionally frozen parameters.

## Acceptance criteria

- [ ] Fully frozen inputs remain frozen after widening and low-rank factorization.
- [ ] Weight-frozen/bias-trainable and weight-trainable/bias-frozen cases preserve their separate roles.
- [ ] Already-trainable inputs still expose trainable transformed weights and receive gradients.
- [ ] Root and replaced-child modes are retained, original modules are untouched, and numerical parity tests continue to pass.

## Validation and completion

Minimal role/mode fixture using existing public APIs:

```python
import torch
from torch import nn
from nnx import low_rank_factorize, widen

base = nn.Linear(4, 4).eval()
base.weight.requires_grad_(False)
base.bias.requires_grad_(True)
factors = low_rank_factorize(base, rank=4)
assert not factors[0].weight.requires_grad
assert not factors[1].weight.requires_grad
assert factors[1].bias.requires_grad
assert not factors.training and all(not m.training for m in factors)
net = nn.Sequential(nn.Linear(4, 4), nn.ReLU(), nn.Linear(4, 2)).eval()
net.requires_grad_(False)
wider = widen(net, layer_name="0", new_width=7)
assert not any(p.requires_grad for p in wider.parameters())
assert not wider[0].training and not wider[2].training
```

Add `test_surgery_preserves_weight_bias_roles` and `test_surgery_preserves_mixed_module_modes`, parameterizing all four weight/bias flag combinations plus no bias. Give target and downstream layers opposing flags to detect accidental propagation from the wrong source. For trainable cases backpropagate a nontrivial loss and assert expected gradients; for frozen cases compare original and transformed trainable-name sets and source flags. Retain full-rank SVD/direct-ReLU widening parity, low-rank approximation expectations and RNG preservation.

From the repository root, use `uv sync --frozen --all-extras` for the CI-equivalent environment. Existing focused coverage (files already exist):

```bash
uv run --frozen pytest tests/test_surgery_widen.py tests/test_surgery_low_rank.py
```

Add the proposed regression cases above to those existing test files, then rerun the same command. Finish with:

```bash
uv run --frozen ruff check src/ tests/ examples/ scripts/
uv run --frozen ruff format --check src/ tests/ examples/ scripts/
uv run --frozen pyright --warnings
```

No new dependency is needed for the implementation described. Preserve Python 3.10 compatibility and default-field omission in existing serialization. These commands are completion requirements for implementation; this specification review does not claim they were run or passed.

## Cascading changes and affected consumers

- **Freeze intent reaches the optimizer.** Follow [freeze/unfreeze](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/finetune/freezing.py) through [widen](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/surgery/widen.py) or [low_rank_factorize](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/surgery/low_rank.py) into [build_param_groups](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/finetune/param_groups.py). New factors inherit the original weight role, while the copied/up-projection bias retains its independent bias role. Build a fresh optimizer only after surgery and inspect membership by Parameter identity: frozen replacement tensors must not silently enter explicit groups. The ordinary ungrouped optimizer may still hold frozen parameters; the invariant there is no gradients or updates, not necessarily physical exclusion.

- **Higher training boundaries.** [NNModel training preflight](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/nn_model.py) rejects a fully frozen network and unsupported low-rank reconstruction topology. Preserving freeze flags must not bypass those guards just to make a new transformed model train. For supported widening, use a correctly rebuilt descriptor and a fresh optimizer; for current low-rank workflows, retain explicit manual refinement. Do not reuse the old optimizer’s stale Parameter objects or claim an exact warm resume across changed topology. `requires_grad` and runtime training modes are not carried by state_dict alone, so preserving them on the returned object does not establish cross-process freeze-policy persistence.

- **Refinement sample with intentional trainability.** [Example 20](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/examples/20_low_rank_surgery_ffn.py) manually fine-tunes the factorized net and prints trainable counts, making it a concrete downstream consumer of accidental unfreezing. Add an isolated weight-frozen/bias-trainable helper, factorize a layer, build explicit groups and take one update. Assert only the intended trainable role changes while factors, source model and original optimizer remain untouched. Update the surgery guide to state that transforms preserve trainability; users who want newly trainable factors must explicitly unfreeze them before creating their optimizer. Preserve mixed child modes through the same workflow.

### Samples and executable coverage

- [`examples/20_low_rank_surgery_ffn.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/examples/20_low_rank_surgery_ffn.py): Add a bounded frozen-weight/trainable-bias factorization helper that constructs fresh explicit groups, performs one update and checks unchanged factors/source plus changed intended bias; keep full manual refinement separate. **Verify:** Proposed execution node: uv run --frozen pytest tests/test_examples_smoke.py -k surgery_freeze_roles. CPU, offline, temporary working directory; call the bounded helper rather than the full training main. Retain import-without-main coverage and add the callable to explicit execution coverage; import success alone is insufficient.

Update `examples/README.md` with each affected sample’s purpose, extras, input/resource needs and bounded invocation. Preserve import-without-main behavior; add explicit import and execution registration in `tests/test_examples_smoke.py` for new unnumbered scripts. Existing discovery covers only `[0-9][0-9]_*.py`, and only selected scripts execute by default.

### Repository, site and wiki propagation

Edit these canonical sources; the two other columns are generated projections, not independently maintained pages. Proposed guide slugs below are explicit suggestions: reuse the prerequisite’s registered slug when it already exists.

| Canonical repository source | Generated site Markdown | Generated wiki Markdown | Required content change |
|---|---|---|---|
| [`docs/surgery.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/docs/surgery.md) | `generated/site/Model-surgery.md` | `generated/wiki/Model-surgery.md` | Document preservation of weight/bias trainability and mixed modes, explicit unfreeze-before-new-optimizer workflow, supported widening reconstruction versus manual low-rank refinement, and no old-optimizer resume guarantee. |
| [`docs/concepts.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/docs/concepts.md) | `generated/site/Concepts.md` | `generated/wiki/Concepts.md` | Connect freeze/unfreeze to post-surgery param-group construction; distinguish explicit-group exclusion from frozen parameters held by a default optimizer and from state_dict persistence. |

Update affected public docstrings/API catalog and regenerate `docs/api.md`; add the scoped changelog entry and reconcile changed top-level claims. Register new guides in `docs/manifest.yaml` with unique slugs/numbers and consistent headings/navigation. Repository-link validation uses Git-tracked paths, so newly linked source files must be tracked/staged before that check. Generated pages are never edited directly.

After source/API edits, run from the repository root in the existing frozen all-extras environment (regenerate/review the lock first only if this ticket changes dependencies):

```bash
uv run --frozen python -m scripts.docs.build_api_reference
uv run --frozen python -m scripts.docs.build_docs
uv run --frozen mkdocs build --strict
uv run --frozen python -m scripts.docs.build_wiki --check
uv run --frozen pytest tests/test_docs_projection.py tests/test_wiki_projection.py
```

If README changed, run `uv run --frozen python -m scripts.docs.build_pypi_readme` and its `--check` form. If canonical diagram HTML changed, run `uv run --frozen python -m scripts.docs.extract_architecture_svg` before the docs build and verify with `--check`. Compare the changed repository passages with rendered `site/` and wiki projections: values, signatures, limitations, examples, links/anchors and assets must agree. Deterministic generation alone is not a freshness or live-publication check.

**Delivery evidence:** record revision, executed checks and changed page mapping. `.github/workflows/docs.yml` builds PRs/main/develop and publishes Pages/wiki on main pushes. Verify served content after that authorized publication; record deployment failure/skew as pending follow-through. Local rendering is not live publication, and PR preparation requires no manual wiki/site push.

### Additional acceptance criteria — cascading completion

- [ ] A freeze → surgery → fresh explicit param-group → optimizer-update workflow preserves independent weight/bias roles, excludes frozen replacements from explicit groups and updates only the intended trainable tensors.
- [ ] The public NNModel path still rejects fully frozen input and unsupported factorized topology; supported widening refinement uses an accurate new descriptor rather than stale optimizer or checkpoint topology.
- [ ] A default ungrouped optimizer holding frozen transformed tensors produces no gradients or value updates for those tensors, while the original model and any original optimizer continue referencing their unchanged Parameters.
- [ ] The bounded Example 20 frozen-role helper executes its optimizer step and assertions, and the guide requires explicit unfreeze when callers intentionally want trainable replacement factors.
- [ ] The issue-specific changes in `docs/surgery.md`, `docs/concepts.md`, affected API reference, sample catalog and changelog are reflected consistently in repository, generated/rendered site and wiki; documented commands/examples and links pass the checks above, and delivery evidence distinguishes the reviewed revision from any pending live publication.

**Coordination:** Propagation belongs to this fix and existing APIs. Related tickets coordinate shared contracts/files; they are not prerequisites for the bounded repaired path.

Additional nonblocking consumers/coordination: [FIX-005](https://github.com/thekaveh/NNx/issues/204), [FEAT-016](https://github.com/thekaveh/NNx/issues/238).

## Delivery order and dependencies

- **Execution order:** 27 / 67 (priority-aware topological order).
- **Dependency wave:** 1. Items in the same wave may proceed independently when their prerequisites are satisfied.
- **Blocked by:** None within this backlog.
- **Related / optional integration:** [FEAT-016](https://github.com/thekaveh/NNx/issues/238). These links do not block this ticket.

**Dependency rationale.** Independent repair of current behavior. No dependency; FEAT-016 must preserve the chosen replacement policy.

Reviewed against [`0c707eb`](https://github.com/thekaveh/NNx/commit/0c707eb5b80d2173843e60102ba326e4f997749d) on 2026-09-20. Acceptance criteria are prospective; this issue does not claim an implementation or new test pass.

**Comment by thekaveh** [comment]
Resolved.

**Root cause:** `widen` and `low_rank_factorize` rebuild existing Linear layers with `skip_init`-constructed replacements whose parameters default to trainable and whose mode defaults to train. A fully frozen, eval-mode network came back with trainable replacement tensors in train mode (four after widening, three after factorization, where the source had none), so an optimizer built after surgery silently updated previously frozen content.

**What changed**
- New `surgery._utils.copy_param_roles(src, dst)`: each replacement parameter inherits the `requires_grad` flag of the same-named source parameter (weight and bias independently; absent tensors left alone) and the replacement module inherits the source's train/eval mode. Applied after value copy, to freshly built modules only — the caller's Parameters and values are never touched.
- `widen`: the target replacement takes the target's own flags/mode and the downstream replacement takes the downstream layer's own (never cross-propagated), so mixed root/child modes survive the deepcopy. `low_rank_factorize`: both SVD factors inherit the source weight's flag, only `up.bias` inherits the source bias's flag (`down` has no bias), and the returned Sequential and both factors take the source mode. Function-preservation tolerances, `skip_init`/local-RNG behaviour, dtype/device placement, SVD rank/error checks and `deepen`'s genuinely new layers are unchanged.
- Example 20: bounded `surgery_freeze_roles()` (`tests/test_examples_smoke.py -k surgery_freeze_roles`) — weight-frozen / bias-trainable layer → factorize → fresh strict parameter groups → one update; factors frozen and unchanged, only the intended bias and the untouched layer move, source Linear untouched. Docs: `surgery.md` §7 (transforms preserve trainability; `unfreeze` explicitly for trainable factors; flags/modes are not carried by `state_dict`), `concepts.md` surgery bullet, examples catalog, CHANGELOG.

**PRs:** #289 (develop, merged), #290 (main, merged).

**Checks run (both branches):**
- `uv run --frozen pytest -q` → 1702 passed, 1 skipped, 2 deselected (the two libcairo-dependent `test_docs_projection` tests — pre-existing local-env failures, unrelated; CI runs them and passed). New: `test_surgery_preserves_weight_bias_roles[…]` (4 flag combos + no-bias ×2, opposing target/downstream flags), `test_surgery_preserves_mixed_module_modes`, `test_widen_frozen_network_cannot_enter_explicit_param_groups`, `test_low_rank_preserves_weight_bias_roles[…]`, `test_low_rank_preserves_module_mode[train|eval]`; existing parity / RNG / SVD tests unchanged.
- `ruff check` → All checks passed; `ruff format --check` → all formatted; `pyright --warnings` → 0 errors, 0 warnings
- docs chain (`build_api_reference` / `build_docs` / `mkdocs build --strict` / `build_wiki --check` / projection tests) → pass; `python examples/20_low_rank_surgery_ffn.py` runs end to end.
- CI: lint-and-test on Python 3.10–3.14, pip-audit, CodeQL, docs build all green on #289 and #290.

**Not verifiable locally:** none. Live Pages/wiki publication is left to the `docs.yml` main-push job.
